storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java

```diff
@@ -0,0 +1,769 @@
+
+                if (LOG.isDebugEnabled() || dynamicState.state != nextState.state) {
+                    LOG.info("STATE {} -> {}", dynamicState, nextState);
+                }
+                //Save the current state for recovery
+                if (!equivilant(nextState.currentAssignment, dynamicState.currentAssignment)) {
+                    LOG.warn("SLOT {}: Changing current assignment from {} to {}", staticState.port, dynamicState.currentAssignment, nextState.currentAssignment);
+                    synchronized(staticState.localState) {
+                        Map<Integer, LocalAssignment> assignments = staticState.localState.getLocalAssignmentsMap();
+                        if (assignments == null) {
+                            assignments = new HashMap<>();
+                        }
+                        if (nextState.currentAssignment == null) {
+                            assignments.remove(staticState.port);
+                        } else {
+                            assignments.put(staticState.port, nextState.currentAssignment);
+                        }
+                        staticState.localState.setLocalAssignmentsMap(assignments);
+                    }
+                    Map<Long, LocalAssignment> update = null;
+                    Map<Long, LocalAssignment> orig = null;
+                    do {
+                        Long lport = new Long(staticState.port);
+                        orig = cachedCurrentAssignmants.get();
+                        update = new HashMap<>(orig);
+                        if (nextState.currentAssignment == null) {
+                            update.remove(lport);
+                        } else {
+                            update.put(lport, nextState.currentAssignment);
+                        }
+                    } while (!cachedCurrentAssignmants.compareAndSet(orig, update));
+                }
+                
+                // clean up the profiler actions that are not being processed
+                removed.removeAll(dynamicState.profileActions);
+                removed.removeAll(dynamicState.pendingStopProfileActions);
+                for (TopoProfileAction action: removed) {
+                    try {
+                        clusterState.deleteTopologyProfileRequests(action.topoId, action.request);
+                    } catch (Exception e) {
+                        LOG.error("Error trying to remove profiling request, it will be retried", e);
+                    }
+                }
+                Set<TopoProfileAction> orig, copy;
+                do {
+                    orig = profiling.get();
+                    copy = new HashSet<>(orig);
+                    copy.removeAll(removed);
+                } while (!profiling.compareAndSet(orig, copy));
+                dynamicState = nextState;
+                newAssignment.compareAndSet(localNewAssignment, dynamicState.newAssignment);
```

Comment:
I don't really understand this assignment. The localNewAssignment set in L726 is in L731 put into a DynamicState. The resulting state from doing a step in the state machine is then used to overwrite dynamicState in L759. I can't find anywhere dynamicState.newAssignment is changed in the state machine, so I don't understand how localNewAssignment and dynamicState.newAssignment can be different. Where/when does nextState have its newAssignment changed?